### PR TITLE
fix collect tipsets error, add test case

### LINKED ISSUE #3027: CollectTipSetOfHeightAtLeast new OBO error
### Description
Because of the recent fix in #3026 we are now missing one tipset in the ancestor subchain in the case when the sample height falls on a null block.  This bug should only be an issue in the case a PoSt is submitted at the very last possible block.  In that case validators will incorrectly fail with ["sample height out of range"](https://github.com/filecoin-project/go-filecoin/pull/3026/files#diff-0169262ea0c745894301dc43c1446163R56)

### Acceptance criteria
- Modify `CollectTipSetsOfHeightAtLeast` to handle this OBO error
- Simplify the `GetRecentAncestors` code which has some complexity that will go away when making this change.  Specifically we no longer need two iterators because `CollectTipSetsOfHeightAtLeast` will no longer drop a value needed by `CollectAtMostNTipSets`
- Test this edge case

### PR DESCRIPTION
**Pull request by deaswang**
fixes #3027

### CODE REVIEW

**Comment by anorth** [comment]
@ZenGround0 could you please shepherd this one?

File: chain/get_ancestors.go [hunk 1]
```diff
@@ -51,38 +51,14 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
...
-	}
-	firstExtraRandomnessAncestorsCids, err := provingPeriodAncestors[len(provingPeriodAncestors)-1].Parents()
-	if err != nil {
-		return nil, err
-	}
-	// no parents means hit genesis so return the whole chain
-	if firstExtraRandomnessAncestorsCids.Len() == 0 {
-		return provingPeriodAncestors, nil
-	}
+	return CollectRecentTipSets(ctx, iterator, earliestAncestorHeight, lookback)
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: Blocking: please don't merge together these two functions.  When I wrote "we no longer need two iterators" I didn't mean we don't want two functions operating over the iterator, I meant we don't need two calls to `IterAncestors`.  

We could have done this merging together of these functions before the change in #3026 (in fact I originally did this) but separated out the two functions to help readability and reusability.

Comment (deaswang) [reply chain/get_ancestors.go]: @ZenGround0 separated them. and rebase.

File: chain/get_ancestors.go [hunk 2]
```diff
@@ -51,28 +51,12 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
 		earliestAncestorHeight = types.NewBlockHeight(uint64(0))
 	}
 
-	// Step 1 -- gather all tipsets with a height greater than the earliest
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: Please keep this comment for readability and update it to be accurate: "gather all tipsets up to the first tipset with a height less than or equal to the earliest possible proving period start"

File: chain/get_ancestors.go [hunk 3]
```diff
@@ -51,28 +51,12 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
...
-	firstExtraRandomnessAncestorsCids, err := provingPeriodAncestors[len(provingPeriodAncestors)-1].Parents()
-	if err != nil {
-		return nil, err
-	}
-	// no parents means hit genesis so return the whole chain
-	if firstExtraRandomnessAncestorsCids.Len() == 0 {
-		return provingPeriodAncestors, nil
-	}
 
-	// Step 2 -- gather the lookback tipsets directly preceding provingPeriodAncestors.
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: Please keep this comment

File: chain/get_ancestors.go [hunk 4]
```diff
@@ -51,28 +51,12 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
...
-	lookBackTS, err := provider.GetTipSet(firstExtraRandomnessAncestorsCids)
-	if err != nil {
-		return nil, err
-	}
-	iterator = IterAncestors(ctx, provider, lookBackTS)
 	extraRandomnessAncestors, err := CollectAtMostNTipSets(ctx, iterator, lookback)
 	if err != nil {
 		return nil, err
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: Please update the function comment for `CollectTipSetsOfHeightAtLeast` to accurately reflect the new behavior.

File: chain/get_ancestors.go [hunk 5]
```diff
@@ -99,6 +83,7 @@ func CollectTipSetsOfHeightAtLeast(ctx context.Context, iterator *TipsetIterator
 		}
 		ret = append(ret, iterator.Value())
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: Blocking: Hey @deaswang in order to close out #3027 this function needs a behavior change.  It looks like the original issue wasn't clear about this.  I've updated the issue to [explain this in better detail](https://github.com/filecoin-project/go-filecoin/issues/3027#issuecomment-523106427).  Please follow up with questions here or there if it is still unclear.

File: chain/get_ancestors.go [hunk 6]
```diff
@@ -51,28 +51,12 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
...
-		return provingPeriodAncestors, nil
-	}
 
-	// Step 2 -- gather the lookback tipsets directly preceding provingPeriodAncestors.
-	lookBackTS, err := provider.GetTipSet(firstExtraRandomnessAncestorsCids)
-	if err != nil {
-		return nil, err
-	}
-	iterator = IterAncestors(ctx, provider, lookBackTS)
 	extraRandomnessAncestors, err := CollectAtMostNTipSets(ctx, iterator, lookback)
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: Blocking: @deaswang without the behavior change discussed [below](https://github.com/filecoin-project/go-filecoin/pull/3250/files#r315803061) this change will skip a tipset in the return array which will cause rare errors which will halt the chain's progress.

Once you implement the described behavior change to `CollectTipSetsOfHeightAtLeast` then this change to `GetRecentAncestors` will be correct.

**Comment by deaswang** [comment]
the requested change have update. @ZenGround0

File: chain/get_ancestors.go [hunk 7]
```diff
@@ -94,11 +89,12 @@ func CollectTipSetsOfHeightAtLeast(ctx context.Context, iterator *TipsetIterator
 		if err != nil {
 			return nil, err
 		}
-		if types.NewBlockHeight(h).LessThan(minHeight) {
-			return ret, nil
-		}
 		ret = append(ret, iterator.Value())
+		if types.NewBlockHeight(h).LessEqual(minHeight) {
+			break
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: 👍

File: chain/get_ancestors.go [hunk 8]
```diff
@@ -51,37 +51,32 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
...
-	// no parents means hit genesis so return the whole chain
-	if firstExtraRandomnessAncestorsCids.Len() == 0 {
-		return provingPeriodAncestors, nil
-	}
 
-	// Step 2 -- gather the lookback tipsets directly preceding provingPeriodAncestors.
-	lookBackTS, err := provider.GetTipSet(firstExtraRandomnessAncestorsCids)
-	if err != nil {
-		return nil, err
+	// Step 2 -- skip the next tipset which added if iterator not complete
```
Comment (ZenGround0) [inline chain/get_ancestors.go]: nit: it would be helpful to explain that the iterator is currently pointing at the lowest height element of `provingPeriodAncestors` and needs to be updated to avoid duplicates.

File: chain/get_ancestors.go [hunk 9]
```diff
@@ -51,37 +51,32 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, provider TipSetP
...
 	if err != nil {
 		return nil, err
 	}
 	return append(provingPeriodAncestors, extraRandomnessAncestors...), nil
 }
 
-// CollectTipSetsOfHeightAtLeast collects all tipsets with a height greater
-// than or equal to minHeight from the input tipset.
+// CollectTipSetsOfHeightAtLeast collects all tipsets up to the first tipset with a height less than
+// or equal to the earliest possible proving period start
```
Comment (anorth) [inline chain/get_ancestors.go]: "up to and including"

Two blocking issues:
- The name of this function is now misleading. Maybe `CollectTipSetsPastHeight`? Not sure - it's hard to name this behaviour
- This function has another caller in `chain/traversal.go` and this behaviour changes appears to be incompatible with its needs (excluding the equal-height block). Unfortunately, `CollectTipsToCommonAncestor` does not have direct unit tests.

Some suggestions for resolution:
- Leave this function alone and write more code in `GetRecentAncestors` to traverse the additional block
- Leave this function alone and write a new one for the needs of `GetRecentAncestors`
- Revert this change, then replace `minheight` with a predicate and rename to `CollectTipSetsWhile`.

**Comment by deaswang** [comment]
rebased and test case passed at local. @anorth @ZenGround0
